File: packages/sdk-js/src/resources/pipeline.ts
import { BaseResource } from './base';
import { AxiosRequestConfig } from 'axios';

// The pipeline engine drives every approved configuration canvas through
// deploy -> monitor -> drift-detect. Deployments are created by deploying a
// canvas (there is no standalone "create deployment" endpoint); every other
// operation acts on an existing canvas or deployment.
//
// Mounted at /api/pipeline on the Community Edition server.

export class PipelineResource extends BaseResource {
  protected readonly RESOURCE_PATH = 'pipeline';

  /** Run the app validator against a canvas. POST /api/pipeline/canvas/{canvasId}/validate */
  async validateCanvas(canvasId: string, config?: AxiosRequestConfig): Promise<any> {
    return this.httpClient.post<any>(`${this.RESOURCE_PATH}/canvas/${canvasId}/validate`, undefined, config);
  }

  /**
   * Queue a deployment of an approved canvas to a target environment.
   * POST /api/pipeline/canvas/{canvasId}/deploy — body: { environmentId, strategy? }.
   */
  async deployCanvas(
    canvasId: string,
    data: { environmentId: string; strategy?: string; [key: string]: any },
    config?: AxiosRequestConfig,
  ): Promise<any> {
    return this.httpClient.post<any>(`${this.RESOURCE_PATH}/canvas/${canvasId}/deploy`, data, config);
  }

  /** Deployment history for a canvas. GET /api/pipeline/canvas/{canvasId}/deployments */
  async listCanvasDeployments(canvasId: string, params?: Record<string, any>, config?: AxiosRequestConfig): Promise<any[]> {
    return this.httpClient.get<any[]>(`${this.RESOURCE_PATH}/canvas/${canvasId}/deployments`, { ...config, params });
  }

  /** Detailed status of a deployment (including logs). GET /api/pipeline/deployments/{deploymentId} */
  async getDeployment(deploymentId: string, config?: AxiosRequestConfig): Promise<any> {
    return this.httpClient.get<any>(`${this.RESOURCE_PATH}/deployments/${deploymentId}`, config);
  }

  /**
   * Roll a deployment back to the previous version.
   * POST /api/pipeline/deployments/{deploymentId}/rollback — body: { reason }.
   */
  async rollbackDeployment(
    deploymentId: string,
    data: { reason: string; [key: string]: any },
    config?: AxiosRequestConfig,
  ): Promise<any> {
    return this.httpClient.post<any>(`${this.RESOURCE_PATH}/deployments/${deploymentId}/rollback`, data, config);
  }

  /** Pause an in-progress (canary/rolling) deployment. POST /api/pipeline/deployments/{deploymentId}/pause */
  async pauseDeployment(deploymentId: string, config?: AxiosRequestConfig): Promise<any> {
    return this.httpClient.post<any>(`${this.RESOURCE_PATH}/deployments/${deploymentId}/pause`, undefined, config);
  }

  /** Resume a paused deployment. POST /api/pipeline/deployments/{deploymentId}/resume */
  async resumeDeployment(deploymentId: string, config?: AxiosRequestConfig): Promise<any> {
    return this.httpClient.post<any>(`${this.RESOURCE_PATH}/deployments/${deploymentId}/resume`, undefined, config);
  }

  /**
   * Promote a successful deployment to the next environment.
   * POST /api/pipeline/deployments/{deploymentId}/promote — body: { targetEnvironmentId }.
   */
  async promoteDeployment(
    deploymentId: string,
    data: { targetEnvironmentId: string; [key: string]: any },
    config?: AxiosRequestConfig,
  ): Promise<any> {
    return this.httpClient.post<any>(`${this.RESOURCE_PATH}/deployments/${deploymentId}/promote`, data, config);
  }

  /** Pipeline dashboard metrics for the customer. GET /api/pipeline/summary */
  async getSummary(config?: AxiosRequestConfig): Promise<any> {
    return this.httpClient.get<any>(`${this.RESOURCE_PATH}/summary`, config);
  }

  /** Matrix of canvases and their deployment status per environment. GET /api/pipeline/environment-matrix */
  async getEnvironmentMatrix(config?: AxiosRequestConfig): Promise<any> {
    return this.httpClient.get<any>(`${this.RESOURCE_PATH}/environment-matrix`, config);
  }

  /** Drift-detection records for the customer. GET /api/pipeline/drift */
  async listDrift(params?: Record<string, any>, config?: AxiosRequestConfig): Promise<any> {
    return this.httpClient.get<any>(`${this.RESOURCE_PATH}/drift`, { ...config, params });
  }

  /**
   * Acknowledge and resolve a drift record.
   * POST /api/pipeline/drift/{driftId}/resolve — body: { action }.
   */
  async resolveDrift(
    driftId: string,
    data: { action: string; [key: string]: any },
    config?: AxiosRequestConfig,
  ): Promise<any> {
    return this.httpClient.post<any>(`${this.RESOURCE_PATH}/drift/${driftId}/resolve`, data, config);
  }
}

File: packages/sdk-python/veltrix_sdk/resources/pipeline.py
# -*- coding: utf-8 -*-

from .base import BaseResource


class PipelineResource(BaseResource):
    """Handles the deployment pipeline (``/api/pipeline``).

    The pipeline engine drives every approved configuration canvas through
    deploy -> monitor -> drift-detect. Deployments are created by deploying a
    canvas (there is no standalone "create deployment" endpoint); every other
    operation acts on an existing canvas or deployment.
    """
    RESOURCE_PATH = "pipeline"

    def validate_canvas(self, canvas_id, **kwargs):
        """Runs the app validator against a canvas (POST /canvas/{id}/validate)."""
        return self._http_client.post(f"{self.RESOURCE_PATH}/canvas/{canvas_id}/validate", **kwargs)

    def deploy_canvas(self, canvas_id, data, **kwargs):
        """Queues a deployment of an approved canvas to a target environment.

        POST /canvas/{id}/deploy -- ``data`` requires ``environmentId`` and
        accepts an optional ``strategy``.
        """
        return self._http_client.post(f"{self.RESOURCE_PATH}/canvas/{canvas_id}/deploy", data=data, **kwargs)

    def list_canvas_deployments(self, canvas_id, params=None, **kwargs):
        """Deployment history for a canvas (GET /canvas/{id}/deployments)."""
        return self._http_client.get(f"{self.RESOURCE_PATH}/canvas/{canvas_id}/deployments", params=params, **kwargs)

    def get_deployment(self, deployment_id, **kwargs):
        """Detailed status of a deployment (GET /deployments/{id})."""
        return self._http_client.get(f"{self.RESOURCE_PATH}/deployments/{deployment_id}", **kwargs)

    def rollback_deployment(self, deployment_id, data, **kwargs):
        """Rolls a deployment back to the previous version.

        POST /deployments/{id}/rollback -- ``data`` requires ``reason``.
        """
        return self._http_client.post(f"{self.RESOURCE_PATH}/deployments/{deployment_id}/rollback", data=data, **kwargs)

    def pause_deployment(self, deployment_id, **kwargs):
        """Pauses an in-progress (canary/rolling) deployment (POST /deployments/{id}/pause)."""
        return self._http_client.post(f"{self.RESOURCE_PATH}/deployments/{deployment_id}/pause", **kwargs)

    def resume_deployment(self, deployment_id, **kwargs):
        """Resumes a paused deployment (POST /deployments/{id}/resume)."""
        return self._http_client.post(f"{self.RESOURCE_PATH}/deployments/{deployment_id}/resume", **kwargs)

    def promote_deployment(self, deployment_id, data, **kwargs):
        """Promotes a successful deployment to the next environment.

        POST /deployments/{id}/promote -- ``data`` requires ``targetEnvironmentId``.
        """
        return self._http_client.post(f"{self.RESOURCE_PATH}/deployments/{deployment_id}/promote", data=data, **kwargs)

    def get_summary(self, **kwargs):
        """Pipeline dashboard metrics for the customer (GET /summary)."""
        return self._http_client.get(f"{self.RESOURCE_PATH}/summary", **kwargs)

    def get_environment_matrix(self, **kwargs):
        """Matrix of canvases and their deployment status per environment (GET /environment-matrix)."""
        return self._http_client.get(f"{self.RESOURCE_PATH}/environment-matrix", **kwargs)

    def list_drift(self, params=None, **kwargs):
        """Drift-detection records for the customer (GET /drift)."""
        return self._http_client.get(f"{self.RESOURCE_PATH}/drift", params=params, **kwargs)

    def resolve_drift(self, drift_id, data, **kwargs):
        """Acknowledges and resolves a drift record.

        POST /drift/{id}/resolve -- ``data`` requires ``action``.
        """
        return self._http_client.post(f"{self.RESOURCE_PATH}/drift/{drift_id}/resolve", data=data, **kwargs)
